packages/formstr-app/src/containers/CreateFormNew/components/FormDetails/index.tsx

```diff
@@ -84,53 +84,99 @@ export const FormDetails: React.FC<FormDetailsProps> = ({
-    if (viewKey) secrets = `${secrets}:${viewKey}`;
-    forms.push(["f", `${key}`, `${relay}`, `${secrets}`]);
-    let encryptedString = await window.nostr.nip44.encrypt(
-      userPub,
-      JSON.stringify(forms)
-    );
-    let myFormEvent: UnsignedEvent = {
-      kind: KINDS.myFormsList,
-      content: encryptedString,
-      pubkey: userPub,
-      tags: [],
-      created_at: Math.round(Date.now() / 1000),
-    };
-    let signedEvent = await window.nostr.signEvent(myFormEvent);
-    await Promise.allSettled(pool.publish(getDefaultRelays(), signedEvent));
-    setSavedOnNostr("saved");
+    
+    saveToMyForms.isRunning = true;
+    setSavedOnNostr("saving");
+  
+    const pool = new SimplePool();
+    const abortController = new AbortController();
+    const relays = getDefaultRelays();
+    let timeoutId: ReturnType<typeof setTimeout> | undefined;
+  
+    try {
+      const timeoutPromise = new Promise((_, reject) => {
+        timeoutId = setTimeout(() => {
+          console.log('Operation timed out after 15s');
+          reject(new Error('Operation timed out'));
+        }, 15000);
+      });
+  
+      const operationPromise = async () => {
+        try {
+          let existingList = await pool.querySync(
+            relays,
+            { kinds: [KINDS.myFormsList], authors: [userPub] }
+          );
+  
+          let forms: Tag[] = [];
+          if (existingList[0]) {
+            let formsString = await window.nostr.nip44.decrypt(
+              userPub,
+              existingList[0].content
+            );
+            forms = JSON.parse(formsString);
+          }
+  
+          let key = `${formAuthorPub}:${formId}`;
+          if (forms.map((f) => f[1]).includes(key)) {
+            console.log('Form already exists in your saved forms');
+            return "already_saved";
+          }
+  
+          let secrets = `${formAuthorSecret}`;
+          if (viewKey) secrets = `${secrets}:${viewKey}`;
+          forms.push(["f", key, relay, secrets]);
+  
+          let encryptedString = await window.nostr.nip44.encrypt(
+            userPub,
+            JSON.stringify(forms)
+          );
+          
+          let myFormEvent: UnsignedEvent = {
+            kind: KINDS.myFormsList,
+            content: encryptedString,
+            pubkey: userPub,
+            tags: [],
+            created_at: Math.round(Date.now() / 1000),
+          };
+  
+          const signedEvent = await window.nostr.signEvent(myFormEvent);
+          await Promise.allSettled(
+            pool.publish(relays, signedEvent)
+          );
+  
+          return "saved";
+        } catch (error) {
+          console.error('Save operation failed:', error);
+          throw error;
+        }
+      };
+  
+      const result = await Promise.race([
+        operationPromise(),
+        timeoutPromise
+      ]);
+  
+      if (timeoutId) clearTimeout(timeoutId);
+      setSavedOnNostr(result === "already_saved" ? "saved" : "saved");
```

Comment:
why do we need a seperate state for already_saved? if it esssentially boils down to the same state?

Reply:
yes, we can condense into a single one